safe_control_gym/experiment.py

```diff
@@ -1,148 +1,238 @@
             print("{}: {:.3f}".format(colored(metric_key,"yellow"), metric_val))
         print("Evaluation done.")
-        return results
+        return trajs_data, metrics
     
-    def collect_trajs(self, ctrl, env, log_freq=None, n_trajs=None, n_steps=None, **kwargs):
-        """"""
-        if not is_wrapped(env, RecordDataWrapper):
-            env = RecordDataWrapper(env)
+    def _execute_evaluations(self, n_episodes=10, n_steps=None, log_freq=None, **kwargs):
+        """Runs the experiments and collects all the required data.
+
+        Args:     
+            n_episodes (int): number of runs to execute
+            n_steps (int): the number of steps to collect in total
+            log_freq (int): the frequency with which to log information
+        
+        Returns:
+            trajs_data (defaultdict(list)): the raw data from the executed runs
+        """
+
+        if n_episodes is None and n_steps is None:
+            raise ValueError('One of n_episodes or n_steps must be defined.')
+        elif n_episodes is not None and n_steps is not None:
+            raise ValueError('Only one of n_episodes or n_steps can be defined.')
         
         # initialize
-        sim_steps = log_freq // env.CTRL_FREQ if log_freq else 1 
-        n_step, n_traj = 0, 0
-        done = False
+        sim_steps = log_freq // self.env.CTRL_FREQ if log_freq else 1 
+        steps, trajs = 0, 0
         ctrl_data = defaultdict(list)
-        if env.INFO_IN_RESET:
-            obs, info = env.reset()
+        if self.env.INFO_IN_RESET:
+            obs, info = self.env.reset()
         else:
-            obs = env.reset()
+            obs = self.env.reset()
             info = None
-        # cache anything the ctrl needs for running an episode later
-        ctrl.reset_before_run(obs, info, env=env)
-        for data_key, data_val in ctrl.get_eval_result_dict():
+
+        self.ctrl.reset_before_run(obs, info, env=self.env)
+
+        if n_episodes is not None:
+            while trajs < n_episodes:
+                action = self.ctrl.select_action(obs, info)
+                # inner sim loop to accomodate different control frequencies
+                for _ in range(sim_steps):
+                    obs, reward, done, info = self.env.step(action)
+                    if done:
+                        trajs += 1
+                        self._evaluation_reset(ctrl_data=ctrl_data)
+                        break
+        elif n_steps is not None:
+            while steps < n_steps: 
+                action = self.ctrl.select_action(obs, info)
+                # inner sim loop to accomodate different control frequencies
+                for _ in range(sim_steps):
+                    obs, reward, done, info = self.env.step(action)
+                    steps += 1 
+                    if steps >= n_steps:
+                        self.env.save_data()
+                        for data_key, data_val in self.ctrl.results_dict:
```

Comment:
it should be `self.ctrl.results_dict.items()` here? i think there's another occurrence of the somewhere.

Reply:
Hmm I forgot to test this, I'll test it now